tests/posix/common/src/key.c

```diff
@@ -144,6 +144,42 @@ ZTEST(key, test_correct_key_is_deleted)
 	}
 }
 
+static void *setspecific_thread(void *count)
+{
+	int value = 42;
+	int *alloc_count = count;
+
+	while (1) {
+		pthread_key_t key;
+
+		zassert_ok(pthread_key_create(&key, NULL), "failed to create key");
+		if (pthread_setspecific(key, &value) == ENOMEM) {
+			break;
+		};
+		*alloc_count += 1;
+	}
+
+	return NULL;
+}
+
+ZTEST(key, test_thread_specific_data_deallocation)
+{
+	pthread_t thread;
+	int alloc_count_t0 = 0;
+	int alloc_count_t1 = 0;
```

Comment:
With hardware stack protection, and these counters in stack memory, the thread that allocates keys will likely trigger a fault on write.

I would suggest making the variables static or moving them to the global scope (or both).

```suggestion
	static int alloc_count_t0 = 0;
	static int alloc_count_t1 = 0;
```